compiler/rustc_hir_typeck/src/fn_ctxt/_impl.rs

```diff
@@ -1351,7 +1355,16 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         let ty = tcx.type_of(def_id);
         assert!(!args.has_escaping_bound_vars());
         assert!(!ty.skip_binder().has_escaping_bound_vars());
-        let ty_instantiated = self.normalize(span, ty.instantiate(tcx, args));
+
+        let ty_instantiated = match res {
+            // types of const parameters are somewhat special as they are part of
+            // the same environment as the const parameter itself. this means that
+            // unlike most paths `type-of(N)` can return a type naming parameters
+            // introduced by the containing item, rather than provided through `N`.
+            Res::Def(DefKind::ConstParam, _) => ty.instantiate_identity(),
+            _ => ty.instantiate(tcx, args),
```

Comment:
this is very subtle, please:
- add/link to an example where this would go wrong
- exhaustively match on all expected `Def`s